Домашняя работа 3

Zalberix opened the pull request:
Добавлена функция с для частотного анализа (топ 10)
Добавлены дополнительные тесты

Zalberix committed:
Домашняя работа 3:
Добавлена функция с для частотного анализа (топ 10)

hungry-honeybadger commented on hw03_frequency_analysis/top.go:
```diff
@@ -1,6 +1,101 @@
+		}
+	}()
+
+	re := regexp.MustCompile(`^[^а-яА-ЯёЁa-zA-Z0-9]+|[^а-яА-ЯёЁa-zA-Z0-9]+$`)
```
Регулярное выражение не меняется при каждом вызове функции, поэтому нет никакого смысла компилировать регулярно на каждый вызов функции. Предлагаю сделать это один раз на уровне пакета

hungry-honeybadger commented on hw03_frequency_analysis/top.go:
```diff
@@ -1,6 +1,101 @@
+	}
+
+	if len(topFrequencyWord) > 10 {
+		return topFrequencyWord[:10]
```
Очень спорное действие. В go slice это указатель на место в памяти, капасити и лен. В данном случае мы просто говорим что лен=10. Но при этом данные как были на своем месте так и остались. И если предположить что у нас текст это список слова русского языка то мы будет таскать слайс из 200 000 слов, ради 10

Zalberix replied:
Про такое поведение не подумал, добавил копирование в новый слайс

Zalberix committed:
fix: Копиляция регулярного выражения вынесено на уровень пакета, а также убрана обработка паники.
fix: Исправлено возвращение подслайса из функции Top10 в случае если слайс больше 10 и теперь возвращается копия первых 10 элементов.

Zalberix committed:
Merge branch 'master' into hw03_frequency_analysis
